### PR TITLE
Add text generation functionality

### PR DESCRIPTION
**Pull request by norpadon**
# Added text generation functionality.

### Wrapping the `Decoder` in `LanguageModel`

The core of the model is the `Decoder`, but for text generation, you should use the `lalamo.language_model.LanguageModel` class. This class acts as a high-level wrapper around the `Decoder` and manages the complexities of the generation process, such as state management, KV caching, and sampling.

To get started, you first load a model's `Decoder` and then wrap it with `LanguageModel`:

```python
from lalamo.language_model import LanguageModel
from lalamo.model_import import import_model, REPO_TO_MODEL

# Import a pre-trained decoder
decoder = import_model(REPO_TO_MODEL["Qwen/Qwen3-0.6B"]).model

# Wrap it in the LanguageModel for generation
language_model = LanguageModel(decoder)
```

### Eager vs. Streaming Generation

The `LanguageModel` offers two primary methods for generating text:

**1. Whole Sequence Generation (`generate`)**

The `generate` method returns a complete sequence of generated token IDs at once. It is the standard way to generate text and is ideal for tasks where the full output is needed before further processing. It is implemented using `jax.lax.scan` for high performance on accelerators.

```python
# Assuming 'prompt_token_ids' is a JAX array of input token IDs
# and 'tokenizer' is available.

response_token_ids = language_model.generate(
    prompt_token_ids,
    max_output_length=64,
    eos_token_ids=jnp.array([tokenizer.eos_token_id]),
)

# The result is a single array containing all generated tokens.
response_text = tokenizer.decode(response_token_ids)
```

**2. Streaming Generation (`stream`)**

The `stream` method returns a Python iterator that yields one token at a time as it's generated. This is useful for interactive applications, like chatbots, where you want to display the response to the user progressively.

Because it involves a Python-level `for` loop and `yield` statements, the `stream` method **cannot be JIT-compiled**.

```python
token_iterator = language_model.stream(
    prompt_token_ids,
    max_output_length=64,
    eos_token_ids=jnp.array([tokenizer.eos_token_id]),
)

# Consume the iterator to get tokens one by one
generated_tokens = []
for token_id in token_iterator:
    generated_tokens.append(token_id)
    # In a real application, you might decode and display the token here.

response_text = tokenizer.decode(jnp.array(generated_tokens))
```

### Performance Optimizations

JAX provides powerful transformations to optimize execution.

**1. Compiling with `jax.jit`**

You can significantly speed up the `generate` method by applying Just-In-Time (JIT) compilation with `jax.jit`. This compiles the Python code into highly optimized XLA code for your target accelerator (GPU/TPU).

Arguments that control the shape and computation graph of the function, like `max_output_length`, must be marked as static.

```python
from jax import jit

# Create a compiled version of the generate function
jit_generate = jit(
    language_model.generate,
    static_argnames=["max_output_length", "eos_token_ids"],
)

# Use the compiled function
response_token_ids = jit_generate(
    prompt_token_ids,
    max_output_length=64,
    eos_token_ids=jnp.array([tokenizer.eos_token_id]),
)
```

**2. Batching with `jax.vmap`**

To process multiple prompts in a single, efficient batch, you can use `jax.vmap` to vectorize the `generate` function.

When batching, prompts often have different lengths. To handle this, you must pad the tokenized prompts to the same length and, crucially, provide the original, unpadded lengths to the `generate` function via the `prompt_length_without_padding` argument. This ensures that the model's attention mechanism correctly ignores the padding tokens.

```python
from jax import vmap

# Example prompts of different lengths
prompts = ["Hello", "How are you today?"]
tokenized_prompts = [tokenizer.encode(p) for p in prompts]

# Get original lengths
prompt_lengths = jnp.array([len(p) for p in tokenized_prompts])

# Pad to the same length
max_len = max(prompt_lengths)
padded_prompts = jnp.array([
    jnp.pad(p, (0, max_len - len(p))) for p in tokenized_prompts
])

# Define a function to vmap
def batched_generate_fn(tokens, length):
    return language_model.generate(
        tokens,
        prompt_length_without_padding=length,
        max_output_length=64
    )

# Vectorize and JIT-compile the function
compiled_batched_generate = jit(vmap(batched_generate_fn))

# Run the batched generation
batched_response_ids = compiled_batched_generate(padded_prompts, prompt_lengths)
```

### The Role of `KVCacheLayer` Implementations

The KV cache is a critical optimization that stores the computed keys (K) and values (V) from previous tokens, so they don't have to be recomputed at every generation step. There are two implementations of `KVCacheLayer`:

**1. `DynamicKVCacheLayer`**

*   **How it works**: This cache implementation dynamically grows in size by using `jnp.concatenate` to append new keys and values.
*   **When it's used**: It is used internally during a single, eager forward pass of the `Attention` module when no existing cache is provided. This is typical during the initial "prefill" step where the input prompt is processed all at once. It is flexible but less efficient for iterative operations.

**2. `StaticKVCacheLayer`**

*   **How it works**: This cache operates on a pre-allocated buffer of a fixed `capacity`. It updates the buffer in-place using the efficient `jax.lax.dynamic_update_slice_in_dim` function. This is much faster and more JAX-friendly as it avoids changing array shapes, which would trigger re-compilations.
*   **When it's used**: This is the cache used for iterative decoding. The `LanguageModel` automatically creates a `StaticKVCacheLayer` with a sufficient capacity (`prompt_length + max_output_length`) before starting the `generate` or `stream` loop. This ensures that the subsequent token-by-token generation is highly performant.

You don't typically need to interact with the caches directly. The `LanguageModel` abstracts this away, by preallocating a `StaticKVCacheLayer ` with the size reserved for `max_output_length` output tokens.

### CODE REVIEW

File: lalamo/modules/kv_cache.py [hunk 1]
```diff
@@ -1,18 +1,196 @@
...
+
+        result = query_indices[:, None] >= key_indices[None, :]
+        if sliding_window_size is not None:
+            swa_mask = query_indices[:, None] < (key_indices[None, :] + sliding_window_size)
+            result = result & swa_mask
+
+        return result
+
+    @property
+    def padding_mask(self) -> Bool[Array, ""] | None:
```
[the remark this answers is not in the document]
Comment (norpadon) [reply lalamo/modules/kv_cache.py]: Fixed

File: pyproject.toml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 [project]
 name = "lalamo"
 description = "Add your description here"
-requires-python = "~=3.12"
+requires-python = ">=3.12, <4"
```
[the remark this answers is not in the document]
Comment (norpadon) [reply pyproject.toml:4]: It is fine

File: tests/test_generation.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        <think>\n\n</think>\n\n
+    """.lstrip(),
+    )
+    tokens = jnp.array(tokenizer.encode(prompt))
+    return GenerationInput(prompt, tokens)
+
+
+@pytest.fixture
+def language_model() -> LanguageModel:
+    decoder = import_model(REPO_TO_MODEL["Qwen/Qwen3-0.6B"]).model
```
Comment (eugenebokhan) [inline tests/test_generation.py:51]: Maybe this would be more clear?

```suggestion
    decoder = import_model(REPO_TO_MODEL["Qwen/Qwen3-0.6B"]).decoder
```

**Review by eugenebokhan — APPROVED** [review]
